zim/demo-library/demo-generator.py

```diff
@@ -0,0 +1,398 @@
+
+# in-ZIM metadata to in-library XML attributes
+NAMES_MAP: dict[str, str] = {
+    "Title": "title",
+    "Description": "description",
+    "Language": "language",
+    "Creator": "creator",
+    "Publisher": "publisher",
+    "Name": "name",
+    "Flavour": "flavour",
+    "Tags": "tags",
+    "Date": "date",
+}
+
+COPIED_KEYS = ["id", "size", "url", "mediaCount", "articleCount"] + list(
+    NAMES_MAP.values()
+)
+
+
+FILENAME_FMT = re.compile(
+    r"^(?P<project>.+?_)(?P<lang>[a-z\-]{2,10}?_|)"
+    r"(?P<option>.+_|)(?P<year>[\d]{4}|)\-(?P<month>[\d]{2})(?P<version>[a-z]{0,2})$",
+    re.IGNORECASE,
+)
+
+REQUESTS_TIMEOUT = 10
+
+
+# Define Pydantic models
+class Demo(BaseModel):
+    name: str
+    zims: list[str]
+    expired_on: datetime.date
+
+
+class Config(BaseModel):
+    demos: list[Demo]
+
+
+class ZimInfo(BaseModel):
+    path: Path
+    infos: dict[str, Any] = {}
+
+
+def binary_human_size(size: float):
+    return human_size(size, binary=True)
+
+
+def install_demo(demo: Demo, demo_tmpl: Template):
+    demo_path = demos_path / demo.name
+    demo_path.mkdir(exist_ok=True)
+    (demo_path / "index.html").write_text(
+        demo_tmpl.render(
+            {"zims": [zims[zim_archive].infos for zim_archive in demo.zims]}
+        )
+    )
+
+
+# constants
+demos_path = Path("/data/demo")
+normal_zim_path = Path("/data/zim")
+hidden_zim_path = Path("/data/hidden-zim")
+demo_library = demos_path / "demo_library.xml"
+demo_config = demos_path / "demos.yaml"
+demo_dead_html = demos_path / "dead.html"
+demo_favicon = demos_path / "favicon.ico"
+demo_dead_kiwix_png = demos_path / "dead-kiwix.png"
+demo_home_tmpl_path = demos_path / "demo-home.jinja2"
+favicons_path = demos_path / "favicons"
+favicons_path.mkdir(exist_ok=True)
+logger = getLogger("demo-generator")
+assets_base_url = (
+    "https://raw.githubusercontent.com/kiwix/operations/demo-library/zim/demo-library/"
+)
+logger.info(f"Assets will be downloaded from {assets_base_url}")
+varnish_cache_url = "http://demo-library-webcache-service"
+
+
+def zim_date_key(s: str) -> tuple[str, int, str]:
+    """Function to be used as sorting criteria for ZIM names.
+
+    Function properly sort by name then date, including versions suffix in the date
+
+    E.g. sorted(
+        [
+            "tests_eng_test_2024-01aa",
+            "tests_eng_test_2024-01b",
+            "tests_eng_test_2024-01",
+            "tests_eng_test_2023-01",
+        ],
+        key=zim_date_key,
+    ) == [
+        "tests_eng_test_2023-01",
+        "tests_eng_test_2024-01",
+        "tests_eng_test_2024-01b",
+        "tests_eng_test_2024-01aa",
+    ]
+
+    The "trick" is to return first up to the xxxx-xx part, then the length of optional
+    ll suffix, then the suffix itself. Python takes care of the rest (first compare the
+    prefix, then the lenght of the suffix, then the suffix itself)
+    """
+    underscore_pos = s.rfind("_")
```

Comment:
Comment is great. I think using a regex here would make the code more readable though (surprisingly)

```py
fm = re.match(r".+_(?P<period>\d{4}-\d{2})(?P<ll>[a-z]*)$", s)
return (fm.groupdict()["period"], len(fm.groupdict()["ll"]), fm.groupdict()["ll"])
```

Reply:
Good point, thanks ! Actually appropriate regex is a bit different, and I enriched it to support split-ZIMs (not sure it will exist in our storage, but who knows ...)